### PR TITLE
Return a more friendly error message when there is no configuration file

### PR DESCRIPTION
**Pull request by liguozhong**
**What this PR does**:

Due to the capacity problem of prometheus, we are trying to find a long-term storage system. We are currently investigating cortex. After starting from the source code for the first time, we found that the startup log cannot be understood. 

```
GOPATH=/Users/fuling/go #gosetup
/usr/local/go/bin/go build -o /private/var/folders/yg/39mj_zbd46532cfbdh1f8fx40000gp/T/GoLand/___go_build_github_com_cortexproject_cortex_cmd_cortex github.com/cortexproject/cortex/cmd/cortex #gosetup
/private/var/folders/yg/39mj_zbd46532cfbdh1f8fx40000gp/T/GoLand/___go_build_github_com_cortexproject_cortex_cmd_cortex
level=info ts=2023-04-24T02:22:01.29409Z caller=main.go:194 msg="Starting Cortex" version="(version=, branch=, revision=7ef779119d1d2e8bcb63f4099dac3f39979880ac)"
level=warn ts=2023-04-24T02:22:01.294151Z caller=cortex.go:388 msg="skipped registration of custom process metrics collector" err="unsupported platform"
level=info ts=2023-04-24T02:22:01.294515Z caller=server.go:323 http=[::]:80 grpc=[::]:9095 msg="server listening on addresses"
level=error ts=2023-04-24T02:22:01.296044Z caller=log.go:117 msg="error running cortex" err="no s3 endpoint in config file\ngithub.com/thanos-io/objstore/providers/s3.validate\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/vendor/github.com/thanos-io/objstore/providers/s3/s3.go:349\ngithub.com/thanos-io/objstore/providers/s3.NewBucketWithConfig\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/vendor/github.com/thanos-io/objstore/providers/s3/s3.go:219\ngithub.com/cortexproject/cortex/pkg/storage/bucket/s3.NewBucketClient\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/storage/bucket/s3/bucket_client.go:29\ngithub.com/cortexproject/cortex/pkg/storage/bucket.NewClient\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/storage/bucket/client.go:102\ngithub.com/cortexproject/cortex/pkg/ingester.New\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/ingester/ingester.go:608\ngithub.com/cortexproject/cortex/pkg/cortex.(*Cortex).initIngesterService\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/cortex/modules.go:404\ngithub.com/cortexproject/cortex/pkg/util/modules.(*Manager).initModule\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/util/modules/modules.go:106\ngithub.com/cortexproject/cortex/pkg/util/modules.(*Manager).InitModuleServices\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/util/modules/modules.go:78\ngithub.com/cortexproject/cortex/pkg/cortex.(*Cortex).Run\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/cortex/cortex.go:398\nmain.main\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/cmd/cortex/main.go:196\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1594\nfailed to create the bucket client\ngithub.com/cortexproject/cortex/pkg/ingester.New\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/ingester/ingester.go:610\ngithub.com/cortexproject/cortex/pkg/cortex.(*Cortex).initIngesterService\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/cortex/modules.go:404\ngithub.com/cortexproject/cortex/pkg/util/modules.(*Manager).initModule\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/util/modules/modules.go:106\ngithub.com/cortexproject/cortex/pkg/util/modules.(*Manager).InitModuleServices\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/util/modules/modules.go:78\ngithub.com/cortexproject/cortex/pkg/cortex.(*Cortex).Run\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/cortex/cortex.go:398\nmain.main\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/cmd/cortex/main.go:196\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1594\nerror initialising module: ingester-service\ngithub.com/cortexproject/cortex/pkg/util/modules.(*Manager).initModule\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/util/modules/modules.go:108\ngithub.com/cortexproject/cortex/pkg/util/modules.(*Manager).InitModuleServices\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/util/modules/modules.go:78\ngithub.com/cortexproject/cortex/pkg/cortex.(*Cortex).Run\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/pkg/cortex/cortex.go:398\nmain.main\n\t/Users/fuling/go/src/github.com/cortexproject/cortex/cmd/cortex/main.go:196\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:250\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1594"

```
According to the documentation, cortex cannot be started directly, and a configuration file must be required. 
image

So we modified the error log
```log
Error loading config ,config.yaml does not exist, please set config.file for custom config path .for example:-config.file=./docs/configuration/single-process-config-blocks-local.yaml
Process finished with the exit code 1
```

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [ ] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by friedrichg** [comment]
Thanks for your contribution @liguozhong !

I agree using the config.file is easier than the flags. It's more readable for humans. But we can't enforce for next release if it wasn't deprecated before.

If we want to deprecate not using config, we can do that adding a deprecated warning message and letting the program continue without exit. And in the following release we can remove the option of no config.file . it's part of the normal process to prevent breaking users between releases.

**Review by friedrichg — CHANGES_REQUESTED** [review]
Remove os.Exit(1) when config.file is not configured

File: cmd/cortex/main.go [hunk 1]
```diff
@@ -82,6 +83,13 @@ func main() {
 			}
 			os.Exit(1)
 		}
+	} else if len(args) == 0 {
+
+		fmt.Fprintf(os.Stderr, "Error loading config ,config.yaml does not exist, please set config.file for custom config path .for example:-config.file=./docs/configuration/single-process-config-blocks-local.yaml")
```
Comment (friedrich-at-adobe) [inline cmd/cortex/main.go:88]: No arguments is an error. That I can agree. But `config.yaml` is not the default. The default is empty

How about something like the following?
```suggestion
		fmt.Fprintf(os.Stderr, "please set configuration file. For example: -config.file=./docs/configuration/single-process-config-blocks-local.yaml")
```

Comment (liguozhong) [reply cmd/cortex/main.go]: thanks

**Comment by liguozhong** [comment]
> Remove os.Exit(1) when config.file is not configured

done, thanks.

**Review by friedrichg — APPROVED** [review]

**Comment by liguozhong** [comment]
thanks

File: cmd/cortex/main.go [hunk 2]
```diff
@@ -82,6 +83,8 @@ func main() {
 			}
 			os.Exit(1)
 		}
+	} else if len(args) == 0 {
+		fmt.Fprintf(os.Stderr, "please set configuration file. For example: -config.file=./docs/configuration/single-process-config-blocks-local.yaml")
```
Comment (friedrichg) [inline cmd/cortex/main.go:87]:
```suggestion
		fmt.Fprintf(os.Stderr, "please set configuration file. For example: -config.file=./docs/configuration/single-process-config-blocks-local.yaml\n")
		os.Exit(1)
```

**Review by friedrichg — CHANGES_REQUESTED** [review]
Getting back at this. I ran this locally and got this

```
go build -o cortex ./cmd/cortex
./cortex
please set configuration file. For example: -config.file=./docs/configuration/single-process-config-blocks-local.yamllevel=info ts=2023-04-26T14:24:24.004725Z caller=main.go:197 msg="Starting Cortex" version="(version=, branch=, revision=293446ad53e71de43aa39cf1a455b64e5822d19a-modified)"
level=warn ts=2023-04-26T14:24:24.004819Z caller=cortex.go:388 msg="skipped registration of custom process metrics collector" err="unsupported platform"
level=info ts=2023-04-26T14:24:24.005789Z caller=server.go:323 http=[::]:80 grpc=[::]:9095 msg="server listening on addresses"
level=error ts=2023-04-26T14:24:24.007306Z caller=log.go:117 msg="error running cortex" err="no s3 endpoint in config file\ngithub.com/thanos-io/objstore/providers/s3.validate\n\t/Volumes/Source/cortexproject/cortex/vendor/github.com/thanos-io/objstore/providers/s3/s3.go:349\ngithub.com/thanos-io/objstore/providers/s3.NewBucketWithConfig\n\t/Volumes/Source/cortexproject/cortex/vendor/github.com/thanos-io/objstore/providers/s3/s3.go:219\ngithub.com/cortexproject/cortex/pkg/storage/bucket/s3.NewBucketClient\n\t/Volumes/Source/cortexproject/cortex/pkg/storage/bucket/s3/bucket_client.go:29\ngithub.com/cortexproject/cortex/pkg/storage/bucket.NewClient\n\t/Volumes/Source/cortexproject/cortex/pkg/storage/bucket/client.go:102\ngithub.com/cortexproject/cortex/pkg/ingester.New\n\t/Volumes/Source/cortexproject/cortex/pkg/ingester/ingester.go:608\ngithub.com/cortexproject/cortex/pkg/cortex.(*Cortex).initIngesterService\n\t/Volumes/Source/cortexproject/cortex/pkg/cortex/modules.go:404\ngithub.com/cortexproject/cortex/pkg/util/modules.(*Manager).initModule\n\t/Volumes/Source/cortexproject/cortex/pkg/util/modules/modules.go:106\ngithub.com/cortexproject/cortex/pkg/util/modules.(*Manager).InitModuleServices\n\t/Volumes/Source/cortexproject/cortex/pkg/util/modules/modules.go:78\ngithub.com/cortexproject/cortex/pkg/cortex.(*Cortex).Run\n\t/Volumes/Source/cortexproject/cortex/pkg/cortex/cortex.go:398\nmain.main\n\t/Volumes/Source/cortexproject/cortex/cmd/cortex/main.go:199\nruntime.main\n\t/usr/local/Cellar/go@1.19/1.19.8/libexec/src/runtime/proc.go:250\nruntime.goexit\n\t/usr/local/Cellar/go@1.19/1.19.8/libexec/src/runtime/asm_amd64.s:1594\nfailed to create the bucket client\ngithub.com/cortexproject/cortex/pkg/ingester.New\n\t/Volumes/Source/cortexproject/cortex/pkg/ingester/ingester.go:610\ngithub.com/cortexproject/cortex/pkg/cortex.(*Cortex).initIngesterService\n\t/Volumes/Source/cortexproject/cortex/pkg/cortex/modules.go:404\ngithub.com/cortexproject/cortex/pkg/util/modules.(*Manager).initModule\n\t/Volumes/Source/cortexproject/cortex/pkg/util/modules/modules.go:106\ngithub.com/cortexproject/cortex/pkg/util/modules.(*Manager).InitModuleServices\n\t/Volumes/Source/cortexproject/cortex/pkg/util/modules/modules.go:78\ngithub.com/cortexproject/cortex/pkg/cortex.(*Cortex).Run\n\t/Volumes/Source/cortexproject/cortex/pkg/cortex/cortex.go:398\nmain.main\n\t/Volumes/Source/cortexproject/cortex/cmd/cortex/main.go:199\nruntime.main\n\t/usr/local/Cellar/go@1.19/1.19.8/libexec/src/runtime/proc.go:250\nruntime.goexit\n\t/usr/local/Cellar/go@1.19/1.19.8/libexec/src/runtime/asm_amd64.s:1594\nerror initialising module: ingester-service\ngithub.com/cortexproject/cortex/pkg/util/modules.(*Manager).initModule\n\t/Volumes/Source/cortexproject/cortex/pkg/util/modules/modules.go:108\ngithub.com/cortexproject/cortex/pkg/util/modules.(*Manager).InitModuleServices\n\t/Volumes/Source/cortexproject/cortex/pkg/util/modules/modules.go:78\ngithub.com/cortexproject/cortex/pkg/cortex.(*Cortex).Run\n\t/Volumes/Source/cortexproject/cortex/pkg/cortex/cortex.go:398\nmain.main\n\t/Volumes/Source/cortexproject/cortex/cmd/cortex/main.go:199\nruntime.main\n\t/usr/local/Cellar/go@1.19/1.19.8/libexec/src/runtime/proc.go:250\nruntime.goexit\n\t/usr/local/Cellar/go@1.19/1.19.8/libexec/src/runtime/asm_amd64.s:1594"
```

There is an end line missing in the first line and now I think exiting is necessary to avoid confusing the user. Because no args is always wrong, you didn't have that before.

**Comment by liguozhong** [comment]
ci fail,datarace . thanos bug?

image

**Comment by friedrichg** [comment]
> ci fail,datarace . thanos bug?
maybe 😄

**Comment by liguozhong** [comment]
file an issue here.
https://github.com/thanos-io/thanos/issues/6321

**Pull request merged**